roleplay.py

```diff
@@ -8,22 +8,118 @@
+                if len(value) > (64 if _f in self.lim_64 else 1024):
+                    raise ValueError('64' if _f in self.lim_64 else '1024')
+                self.__dict__[_f] = self._name(value) if _f == 'name' else value
+            else:
+                self.__dict__[_f] = 'undefined' if _f in self.mandatory_fields else ''
+
+        @classmethod
+        def blank_bio(cls, author: int, name: str):
+            _t = dict(zip(cls.fields, ['']*15))
+            for f in cls.mandatory_fields:
+                _t[f] = 'undefined'
+
+            _t['author'] = author
+            _t['name'] = cls._name(name)
+
+            return cls(**_t)
+
+        @staticmethod
+        def _name(name: str)->str:
+            """
+            removes trailing/leading whitespace, limits whitespace between words to one space, removes newlines.
+            :param name:
+            :return:
+            """
+            _n = re.sub('\s+', ' ', re.sub('^\s+|\s+$|\n|\r', '', name))
+            if not _n:
+                raise CommandSyntaxError('Empty name provided.')
+            return _n[:64]
+
+        def embed(self, guild, roles)-> Embed:
+            """
+            Generates a pretty discord embed of this role.
+            :param guild: guild that the bio belongs to, for member and role searching
+            :param roles: list of accepted race roles
+            :return:
+            """
+            t_embed = Embed(type="rich", colour=16711680)
+
+            role = find_role(guild, self.race)
+
+            t_embed.title = self.name
+
+            if role and role.id in roles:
+                t_embed.colour = role.colour
+
+            t_member = guild.get_member(self.author)
+            if t_member:
+                t_embed.set_footer(text=f"Character belonging to {t_member.display_name}",
+                                   icon_url=t_member.avatar_url)
+
+            t_embed.description = "```\n" +\
+              '\n'.join([f"{f.capitalize():<7}: {self.__dict__[f]}" for f in self.fields[2:6] if self.__dict__[f]])\
+              + "```\n" +\
+              (f"[Theme song.]({self.theme})\n" if self.theme else '') + \
+              (f"[Extended bio.]({self.link})\n" if self.link else '') + \
+              (f"Owner: {t_member.mention}" if t_member else '')
+
+            if self.image:
+                t_embed.set_image(url=self.image)
+
+            for field in (f for f in self.fields[9:] if self.__dict__[f]):
+                t_embed.add_field(name=field.capitalize(), value=self.__dict__[field])
+
+            return t_embed
+
     async def activate(self):
-        self.bios = self.config_manager.get_plugin_config_file("bios.json")
+        self.bios = self.config_manager.get_plugin_config_file\
+            ("bios.json",
+             json_save_args={'default': lambda o: asdict(o), 'indent': 2, 'ensure_ascii': False},
+             json_load_args={'object_hook': lambda o: self.Bio(**o) if {*o} == {*self.Bio.fields} else o})
```

Comment:
I'd prefer to do something more like `self.Bio(**o) if o.get("__bio__")` rather than comparatively expensive set casts. I understand that this was done for compatibility with the old bios file, but that should be, ideally, handled with a conversion check on activation...

Reply:
This also makes the JSON a bit cleaner for manual editing and simplifies some stuff like dumping and object creation since that'd be a useless "__bio__" field to have in the object. 

Also this is still not entirely compatible with old bios because old bios could skip fields and this one expects all the fields. Should probably go back and write a proper conversion logic.

Reply:
I suppose that's valid. I was just thinking, it'd be nice to have a relatively standardized way of (de)serializing objects in JSON, something like a `__classhint__` field. I dunno, something to think about later I suppose.